### PR TITLE
Database for Runtime domain in Director

### PR DESCRIPTION
**Pull request by crabtree**
**Description**

Changes proposed in this pull request:

- Introduce database implementation for runtime repository
- Allow configuring database connection via environment variables for director component

**Related issue(s)**

See also #78

### CODE REVIEW

File: components/director/internal/domain/runtime/repository_test.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+	rows := sqlmock.NewRows([]string{"id", "tenant_id", "name", "description", "status_condition", "status_timestamp", "auth"}).
+		AddRow(uuid.New().String(), tenantID, "Runtime ABC", "Description for runtime ABC", "INITIAL", time, agentAuthStr).
+		AddRow(uuid.New().String(), tenantID, "Runtime XYZ", "Description for runtime XYZ", "INITIAL", time, agentAuthStr)
+
+	sqlMock.ExpectQuery(`^SELECT (.+) FROM "public"."runtimes" WHERE "tenant_id" = \$1$`).
+		WithArgs(tenantID).
+		WillReturnRows(rows)
+
+	ctx := context.WithValue(context.TODO(), appdb.DatabaseCtxKey, sqlxDB)
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/repository_test.go]: to discuss: where we test transactions handling?

Comment (aszecowka) [reply components/director/internal/domain/runtime/repository_test.go]: it will be really nice to have a helper method for decorating context with "db context"

File: components/director/internal/domain/runtime/repository.go [hunk 2]
```diff
@@ -102,3 +107,157 @@ func (r *inMemoryRepository) findRuntimeNameWithinTenant(tenant, name string) bo
...
+	if err != nil {
+		return nil, err
+	}
+
+	stmt := fmt.Sprintf(`SELECT "id", "tenant_id", "name", "description", "status_condition", "status_timestamp", "auth" FROM "%s"."%s" WHERE "id" = $1 AND "tenant_id" = $2`,
+		runtimeTableSchema,
+		runtimeTable)
+
+	var runtimeEnt Runtime
+	err = db.Get(&runtimeEnt, stmt, id, tenant)
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/repository.go]: db.Get vs db.GetContext

File: components/director/internal/domain/runtime/resolver.go [hunk 3]
```diff
@@ -51,11 +62,23 @@ func (r *Resolver) Runtimes(ctx context.Context, filter []*graphql.LabelFilter,
 		cursor = string(*after)
 	}
 
+	tx, err := r.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = r.ctxvs.WithValue(ctx, appdb.DatabaseCtxKey, tx)
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/resolver.go]: probably we should check if context already has a tx?

File: tests/end-to-end/before-commit.sh [hunk 4]
```diff
@@ -113,16 +113,16 @@ fi
...
+# echo -e "${GREEN}? Checking GraphQL examples${NC}"
+# ./gen-examples.sh
+# genExamplesResult=$?
+
+# if [[ ${genExamplesResult} != 0 ]]
+# 	then
+#     	echo -e "${RED}✗ Checking GraphQL examples${NC}\n$genExamplesResult${NC}"
+#     	exit 1;
+# 	else echo -e "${GREEN}√ Checking GraphQL examples${NC}"
+# fi
```
Comment (aszecowka) [inline tests/end-to-end/before-commit.sh]: add TODO here with a link to the issue

File: tests/end-to-end/director/assertions.go [hunk 5]
```diff
@@ -194,5 +194,4 @@ func assertEventsAPI(t *testing.T, in []*graphql.EventAPIDefinitionInput, actual
 func assertRuntime(t *testing.T, in graphql.RuntimeInput, actual graphql.Runtime) {
 	assert.Equal(t, in.Name, actual.Name)
 	assert.Equal(t, *in.Description, *actual.Description)
-	assert.Equal(t, *in.Labels, actual.Labels)
```
Comment (aszecowka) [inline tests/end-to-end/director/assertions.go]: add TODO

File: tests/end-to-end/director/common_test.go [hunk 6]
```diff
@@ -37,7 +37,7 @@ func newAuthorizedHTTPClient() *http.Client {
 
 func (tc *testContext) RunQuery(ctx context.Context, req *gcli.Request, resp interface{}) error {
 	if req.Header["Tenant"] == nil {
-		req.Header["Tenant"] = []string{"test-end-to-end"}
+		req.Header["Tenant"] = []string{"2a1502ba-aded-11e9-a2a3-2a2ae2dbcce4"}
```
Comment (aszecowka) [inline tests/end-to-end/director/common_test.go]: we need to document somewhere about new requirements for tenant header OR
add some mock translation on the gateway from external tenant to the internal tenant.

File: tests/end-to-end/director/runtime_api_test.go [hunk 7]
```diff
@@ -66,22 +66,6 @@ func TestRuntimeCreateUpdateAndDelete(t *testing.T) {
 			}`, actualRuntime.ID, tc.gqlFieldsProvider.ForRuntime()))
 	err = tc.RunQuery(ctx, getRuntimeReq, &actualRuntime)
 	require.NoError(t, err)
-	assert.Len(t, actualRuntime.Labels, 2)
```
Comment (aszecowka) [inline tests/end-to-end/director/runtime_api_test.go]: maybe we should introduce feature flag that disables right now this part of test?

File: components/director/internal/appdb/appdb.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	return db, err
+}
+
+func waitForDatabase(connString string, retryCount int) (Db, error) {
+	var db *sqlx.DB
+	var err error
+	for ; retryCount > 0; retryCount-- {
+		db, err = sqlx.Open("postgres", connString)
+		if err == nil {
+			break
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/appdb/appdb.go]: Definitely not, will check

File: components/director/internal/domain/runtime/repository.go [hunk 9]
```diff
@@ -1,11 +1,16 @@
...
 
+	"github.com/jmoiron/sqlx"
+	"github.com/kyma-incubator/compass/components/director/internal/appdb"
 	"github.com/kyma-incubator/compass/components/director/internal/labelfilter"
 	"github.com/kyma-incubator/compass/components/director/internal/model"
 	"github.com/kyma-incubator/compass/components/director/pkg/pagination"
+	"github.com/lib/pq"
 )
 
 type inMemoryRepository struct {
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository.go]: I will remove the in-memory repository. I think we do not need it anymore.

File: components/director/internal/domain/runtime/repository.go [hunk 10]
```diff
@@ -102,3 +107,157 @@ func (r *inMemoryRepository) findRuntimeNameWithinTenant(tenant, name string) bo
 	}
 	return false
 }
+
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository.go]: As in-memory repository will be removed, no change is needed.

File: components/director/internal/domain/runtime/repository.go [hunk 11]
```diff
@@ -102,3 +107,157 @@ func (r *inMemoryRepository) findRuntimeNameWithinTenant(tenant, name string) bo
...
+	runtimeTableSchema string = "public"
+)
+
+type pgRepository struct{}
+
+func NewPostgresRepository() *pgRepository {
+	return &pgRepository{}
+}
+
+func (r *pgRepository) GetByID(ctx context.Context, tenant, id string) (*model.Runtime, error) {
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository.go]: Thanks to that, we validate the tenant and runtime relation

File: components/director/internal/domain/runtime/repository_test.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	sqlMock.ExpectQuery(`^SELECT (.+) FROM "public"."runtimes" WHERE "id" = \$1 AND "tenant_id" = \$2$`).
+		WithArgs(runtimeID, tenantID).
+		WillReturnRows(rows)
+
+	ctx := context.WithValue(context.TODO(), appdb.DatabaseCtxKey, sqlxDB)
+
+	pgRepository := runtime.NewPostgresRepository()
+
+	// when
+	modelRuntime, err := pgRepository.GetByID(ctx, tenantID, runtimeID)
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository_test.go]: It's not, leftover from debugging :)

File: components/director/internal/appdb/appdb.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	err = db.Ping()
+
+	return NewDB(db), err
+}
+
+//go:generate mockery -name=Db -output=automock -outpkg=automock -case=underscore
+type Db interface {
+	Begin() (DatabaseTx, error)
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/appdb/appdb.go]: In man.go we also call Close si it has to stay, the Ping can be removed.

File: components/director/internal/appdb/appdb.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tgorgol) [reply components/director/internal/appdb/appdb.go]: The error is actually returned by `sqlx.Ping` while `sqlx.Open` returns no error (hence no retries)

File: components/director/internal/domain/runtime/repository_test.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+	pgRepository := runtime.NewPostgresRepository()
+
+	// when
+	err = pgRepository.Create(ctx, modelRuntime)
+
+	// then
+	assert.NoError(t, err)
+}
+
+func TestPgRepository_Update_ShouldUpdateRuntimeEntityFromValidModel(t *testing.T) {
```
Comment (tgorgol) [inline components/director/internal/domain/runtime/repository_test.go]: I see that there are many edge cases not covered by the tests, do we want to test them as well? (not only in this function)

File: components/director/internal/domain/runtime/entity.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+func EntityFromRuntimeModel(model *model.Runtime) (*Runtime, error) {
+	var nullDescription sql.NullString
+	if model.Description != nil && len(*model.Description) > 0 {
+		nullDescription = sql.NullString{
+			String: *model.Description,
+			Valid:  true,
+		}
+	}
+
+	agentAuthMarshalled, err := json.Marshal(model.AgentAuth)
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/entity.go]: IMO it makes sense if the amount of object-specific columns is limited. In this case, it may grow and we will have lots of nullable columns.

File: components/director/cmd/main.go [hunk 16]
```diff
@@ -30,8 +40,20 @@ func main() {
 
 	configureLogger()
 
+	connString := fmt.Sprintf(connStringf, cfg.DatabaseHost, cfg.DatabasePort, cfg.DatabaseUser, cfg.DatabasePassword, cfg.DatabaseName, cfg.DatabaseSSLMode)
+	db, err := appdb.ConfigureDatabase(connString)
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/cmd/main.go]: Done.

File: components/director/internal/domain/runtime/repository.go [hunk 17]
```diff
@@ -102,3 +107,157 @@ func (r *inMemoryRepository) findRuntimeNameWithinTenant(tenant, name string) bo
...
+		return err
+	}
+
+	stmt := fmt.Sprintf(`DELETE FROM "%s"."%s" WHERE "id" = $1`, runtimeTableSchema, runtimeTable)
+	_, err = db.Exec(stmt, item.ID)
+
+	return err
+}
+
+func dbFromCtx(ctx context.Context) (appdb.DatabaseOp, error) {
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository.go]: Moved to appdb package as publically available func that can be called from different packages

File: components/director/internal/domain/runtime/resolver.go (hunk 3 above)
Comment (pkosiec) [reply components/director/internal/domain/runtime/resolver.go]: Why? There will be no error if we want to overwrite value for a context key. And we rather don't want to pass shared context for multiple resolvers in single HTTP request, right?

File: tests/end-to-end/director/runtime_api_test.go (hunk 7 above)
Comment (pkosiec) [reply tests/end-to-end/director/runtime_api_test.go]: It is removed because of lack of the functionality, right? I would be better to comment it then and add TODO to uncomment.

File: components/director/internal/domain/runtime/entity.go [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	"encoding/json"
+	"time"
+
+	model "github.com/kyma-incubator/compass/components/director/internal/model"
+)
+
+type runtimeStatusCondition string
+
+// Runtime struct represents database entity for Runtime
+type Runtime struct {
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/entity.go]: @pkosiec not really we can do this. Entities contain types related to database logic, like `sql.NullString`. On the domain models, we have just a `*string`. I think it is a good separation. Especially if we have different representation in models and entities (see RuntimeStatus in the model and appropriate fields in the entity).

File: components/director/README.md [hunk 19]
```diff
@@ -12,3 +12,19 @@ go run cmd/main.go
 ```
 
 The GraphQL API playground is available at `localhost:3000`.
+
+## Configuration
```
Comment (aszecowka) [inline components/director/README.md]: Consider adding instruction how to run director locally, without creating Docker image

File: components/director/README.md [hunk 20]
```diff
@@ -12,3 +12,19 @@ go run cmd/main.go
...
+
+The Director binary allows to override some configuration parameters. You can specify following environment variables.
+
+| ENV                         | Default        | Description                                       |
+|-----------------------------|----------------|---------------------------------------------------|
+| APP_ADDRESS                 | 127.0.0.1:3000 | The address and port for the service to listen on |
+| APP_DATABASE_USER           | postgres       | Database username                                 |
```
Comment (aszecowka) [inline components/director/README.md]: in config file you have APP_DB_USER

File: components/director/internal/domain/runtime/resolver.go [hunk 21]
```diff
@@ -51,11 +62,23 @@ func (r *Resolver) Runtimes(ctx context.Context, filter []*graphql.LabelFilter,
...
+
 	runtimesPage, err := r.svc.List(ctx, labelFilter, first, &cursor)
 	if err != nil {
 		return nil, err
 	}
 
+	err = tx.Commit()
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/resolver.go]: please check if we should call rollback explicitly, if so maybe try to implement:
```tx.RollbackUnlessCommitted()```

File: components/director/internal/domain/runtime/entity.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	ID              string         `db:"id"`
+	TenantID        string         `db:"tenant_id"`
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/entity.go]: Will be covered by another issue.

File: components/director/internal/domain/runtime/resolver.go (hunk 21 above)
Comment (crabtree) [reply components/director/internal/domain/runtime/resolver.go]: Will be covered by another issue.

File: components/director/internal/appdb/appdb.go [hunk 23]
```diff
@@ -0,0 +1,88 @@
+package appdb
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/appdb/appdb.go]: after longer discussion, we decided to go with `persistence`

File: components/director/README.md (hunk 19 above)
Comment (crabtree) [reply components/director/README.md]: Readme updated.

File: components/director/internal/domain/runtime/entity.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/entity.go]: Issue #172

File: components/director/internal/domain/runtime/repository.go [hunk 24]
```diff
@@ -1,37 +1,69 @@
 package runtime
 
 import (
+	"context"
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/runtime/repository.go]: Issue #169

File: tests/end-to-end/director/common_test.go (hunk 6 above)
Comment (crabtree) [reply tests/end-to-end/director/common_test.go]: The IDs should be always in UUID format, not string. Issue #172

File: components/director/internal/domain/root_resolver.go [hunk 25]
```diff
@@ -59,12 +62,14 @@ func NewRootResolver() *RootResolver {
...
 
+	appCtx := appcontext.NewAppContext()
+
 	return &RootResolver{
 		app:         application.NewResolver(appSvc, apiSvc, eventAPISvc, docSvc, webhookSvc, appConverter, docConverter, webhookConverter, apiConverter, eventAPIConverter),
 		api:         api.NewResolver(apiSvc, appSvc, apiConverter, authConverter),
 		eventAPI:    eventapi.NewResolver(eventAPISvc, appSvc, eventAPIConverter),
 		doc:         document.NewResolver(docSvc, appSvc, frConverter),
-		runtime:     runtime.NewResolver(runtimeSvc, runtimeConverter),
+		runtime:     runtime.NewResolver(transact, appCtx, runtimeSvc, runtimeConverter),
```
[the remark this answers is not in the document]
Comment (crabtree) [reply components/director/internal/domain/root_resolver.go]: In terms of context, yes it is good practice. In this case, the appCtx is a wrapper around the context.WithValue functions, it is not the context instance.

File: components/director/internal/domain/runtime/resolver.go (hunk 21 above)
Comment (aszecowka) [reply components/director/internal/domain/runtime/resolver.go]: not true, it is already covered :)
